Telegraf crashes when buffer_strategy = "disk" and more than one output plugin of the same type is configured

### Relevant telegraf.conf

```toml
# Configuration for telegraf agent
[agent]
  debug = true
  metric_batch_size = 1000
  metric_buffer_limit = 10000
  flush_interval = "10s"
  quiet = false
  omit_hostname = true
  buffer_strategy = "disk"
  buffer_directory = "var"

###############################################################################
#                            OUTPUT PLUGINS                                   #
###############################################################################

# Configuration for sending metrics to InfluxDB
[[outputs.influxdb]]
  alias = "influxdb-data"
  namedrop = ["telegraf*"]
  urls = ["http://127.0.0.1:8086"]
  database = "telegraf"

[[outputs.influxdb]]
  alias = "influxdb-internal"
  namepass = ["telegraf*"]
  urls = ["http://127.0.0.1:8096"]
  database = "monitoring"
  
###############################################################################
#                            INPUT PLUGINS                                    #
###############################################################################

# Collect statistics about itself
[[inputs.internal]]
  name_prefix = "telegraf_"

# Read metrics about cpu usage
[[inputs.cpu]]
  percpu = true
  totalcpu = true
  collect_cpu_time = false
  report_active = false
  core_tags = false
```


### Logs from Telegraf

```text
2024-09-12T12:46:20Z I! Loading config: .\telegraf.conf
2024-09-12T12:46:20Z W! Using disk buffer strategy for plugin outputs.influxdb, this is an experimental feature
2024-09-12T12:46:20Z W! Using disk buffer strategy for plugin outputs.influxdb, this is an experimental feature
2024-09-12T12:46:20Z I! Starting Telegraf 1.32.0 brought to you by InfluxData the makers of InfluxDB
2024-09-12T12:46:20Z I! Available plugins: 235 inputs, 9 aggregators, 32 processors, 26 parsers, 62 outputs, 5 secret-stores
2024-09-12T12:46:20Z I! Loaded inputs: cpu internal
2024-09-12T12:46:20Z I! Loaded aggregators:
2024-09-12T12:46:20Z I! Loaded processors:
2024-09-12T12:46:20Z I! Loaded secretstores:
2024-09-12T12:46:20Z I! Loaded outputs: influxdb (2x)
2024-09-12T12:46:20Z I! Tags enabled:
2024-09-12T12:46:20Z I! [agent] Config: Interval:10s, Quiet:false, Hostname:"", Flush Interval:10s
2024-09-12T12:46:20Z D! [agent] Initializing plugins
2024-09-12T12:46:20Z D! [agent] Connecting outputs
2024-09-12T12:46:20Z D! [agent] Attempting connection to [outputs.influxdb::influxdb-data]
2024-09-12T12:46:20Z D! [agent] Successfully connected to outputs.influxdb::influxdb-data
2024-09-12T12:46:20Z D! [agent] Attempting connection to [outputs.influxdb::influxdb-internal]
2024-09-12T12:46:20Z D! [agent] Successfully connected to outputs.influxdb::influxdb-internal
2024-09-12T12:46:20Z D! [agent] Starting service inputs
2024-09-12T12:46:30Z D! [outputs.influxdb::influxdb-data] Buffer fullness: 0 metrics
2024-09-12T12:46:30Z D! [outputs.influxdb::influxdb-internal] Wrote batch of 8 metrics in 6.1253ms
2024-09-12T12:46:30Z D! [outputs.influxdb::influxdb-internal] Buffer fullness: 8 metrics
2024-09-12T12:46:40Z D! [outputs.influxdb::influxdb-data] Wrote batch of 9 metrics in 7.721ms
2024-09-12T12:46:40Z D! [outputs.influxdb::influxdb-internal] Wrote batch of 16 metrics in 6.6225ms
2024-09-12T12:46:40Z D! [outputs.influxdb::influxdb-internal] Buffer fullness: 22 metrics
2024-09-12T12:46:40Z D! [outputs.influxdb::influxdb-data] Buffer fullness: 22 metrics
2024-09-12T12:46:50Z E! raw metric data: []
2024-09-12T12:46:50Z E! raw metric data: []
panic: failed to decode metric from bytes: EOF

goroutine 57 [running]:
github.com/influxdata/telegraf/models.(*DiskBuffer).Batch(0xc00176f710, 0x3e8)
        /go/src/github.com/influxdata/telegraf/models/buffer_disk.go:146 +0x52f
github.com/influxdata/telegraf/models.(*RunningOutput).Write(0xc0023b0e70)
        /go/src/github.com/influxdata/telegraf/models/running_output.go:292 +0x3a4
github.com/influxdata/telegraf/agent.(*Agent).flushOnce.func1()
        /go/src/github.com/influxdata/telegraf/agent/agent.go:942 +0x23
created by github.com/influxdata/telegraf/agent.(*Agent).flushOnce in goroutine 32
        /go/src/github.com/influxdata/telegraf/agent/agent.go:941 +0xa6
panic: failed to decode metric from bytes: EOF

goroutine 98 [running]:
github.com/influxdata/telegraf/models.(*DiskBuffer).Batch(0xc00176f7a0, 0x3e8)
        /go/src/github.com/influxdata/telegraf/models/buffer_disk.go:146 +0x52f
github.com/influxdata/telegraf/models.(*RunningOutput).Write(0xc0023b0f20)
        /go/src/github.com/influxdata/telegraf/models/running_output.go:292 +0x3a4
github.com/influxdata/telegraf/agent.(*Agent).flushOnce.func1()
        /go/src/github.com/influxdata/telegraf/agent/agent.go:942 +0x23
created by github.com/influxdata/telegraf/agent.(*Agent).flushOnce in goroutine 33
        /go/src/github.com/influxdata/telegraf/agent/agent.go:941 +0xa6
```


### System info

Telegraf 1.32.0 (replicated in windows and kubernetes)

### Docker

_No response_

### Steps to reproduce

1. Run two different InfluxDB instances
2. Run Telegraf with the provided configuration: buffer_strategy = "disk", two InfluxDB output plugin configured, one input to collect metrics and one internal input plugin to collect internal metrics. 
3. Wait some seconds for the error (the flush interval configured?)

### Expected behavior

Telegraf writes metrics to the configured InfluxDBs without crashes.

### Actual behavior

Telegraf stops working. Telegraf uses a folder with the plugin name to write the buffer data to the disk. If the output plugins are of the same type the concurrent go routines maybe are trying to read/write the same data file inside this folder. Maybe by using the alias as folder name could solve the issue.

### Additional info

_No response_

--- a/docs/CONFIGURATION.md
+++ b/docs/CONFIGURATION.md
@@ -361,7 +361,7 @@ The agent table configures Telegraf and the defaults used across all plugins.
 
 - **buffer_directory**:
   The directory to use when in `disk` buffer mode. Each output plugin will make
-  another subdirectory in this directory with the output plugin's name.
+  another subdirectory in this directory with the output plugin's ID.
 
 ## Plugins
 

--- a/models/buffer.go
+++ b/models/buffer.go
@@ -49,7 +49,7 @@ type BufferStats struct {
 }
 
 // NewBuffer returns a new empty Buffer with the given capacity.
-func NewBuffer(name string, alias string, capacity int, strategy string, path string) (Buffer, error) {
+func NewBuffer(id, name, alias string, capacity int, strategy, path string) (Buffer, error) {
 	registerGob()
 
 	bs := NewBufferStats(name, alias, capacity)
@@ -58,7 +58,7 @@ func NewBuffer(name string, alias string, capacity int, strategy string, path st
 	case "", "memory":
 		return NewMemoryBuffer(capacity, bs)
 	case "disk":
-		return NewDiskBuffer(name, alias, path, bs)
+		return NewDiskBuffer(id, path, bs)
 	}
 	return nil, fmt.Errorf("invalid buffer strategy %q", strategy)
 }

--- a/models/buffer.go
+++ b/models/buffer.go
@@ -58,7 +58,7 @@ func NewBuffer(name string, alias string, capacity int, strategy string, path st
 	case "", "memory":
 		return NewMemoryBuffer(capacity, bs)
 	case "disk":
-		return NewDiskBuffer(name, path, bs)
+		return NewDiskBuffer(name, alias, path, bs)
 	}
 	return nil, fmt.Errorf("invalid buffer strategy %q", strategy)
 }

--- a/models/buffer.go
+++ b/models/buffer.go
@@ -49,7 +49,7 @@ type BufferStats struct {
 }
 
 // NewBuffer returns a new empty Buffer with the given capacity.
-func NewBuffer(id, name, alias string, capacity int, strategy, path string) (Buffer, error) {
+func NewBuffer(name, id, alias string, capacity int, strategy, path string) (Buffer, error) {
 	registerGob()
 
 	bs := NewBufferStats(name, alias, capacity)
@@ -58,7 +58,7 @@ func NewBuffer(id, name, alias string, capacity int, strategy, path string) (Buf
 	case "", "memory":
 		return NewMemoryBuffer(capacity, bs)
 	case "disk":
-		return NewDiskBuffer(id, path, bs)
+		return NewDiskBuffer(name, id, path, bs)
 	}
 	return nil, fmt.Errorf("invalid buffer strategy %q", strategy)
 }

--- a/models/buffer_disk.go
+++ b/models/buffer_disk.go
@@ -33,8 +33,13 @@ type DiskBuffer struct {
 	isEmpty bool
 }
 
-func NewDiskBuffer(name string, path string, stats BufferStats) (*DiskBuffer, error) {
-	filePath := filepath.Join(path, name)
+func NewDiskBuffer(name, alias, path string, stats BufferStats) (*DiskBuffer, error) {
+	pluginName := name
+	if alias != "" {
+		pluginName += "-" + alias
+	}
+
+	filePath := filepath.Join(path, pluginName)
 	walFile, err := wal.Open(filePath, nil)
 	if err != nil {
 		return nil, fmt.Errorf("failed to open wal file: %w", err)

--- a/models/buffer_disk.go
+++ b/models/buffer_disk.go
@@ -33,17 +33,18 @@ type DiskBuffer struct {
 	isEmpty bool
 }
 
-func NewDiskBuffer(name, alias, path string, stats BufferStats) (*DiskBuffer, error) {
-	pluginName := name
-	if alias != "" {
-		pluginName += "-" + alias
-	}
-
-	filePath := filepath.Join(path, pluginName)
+func NewDiskBuffer(id, path string, stats BufferStats) (*DiskBuffer, error) {
+	filePath := filepath.Join(path, id)
 	walFile, err := wal.Open(filePath, nil)
 	if err != nil {
 		return nil, fmt.Errorf("failed to open wal file: %w", err)
 	}
+	//nolint:errcheck // cannot error here
+	if index, _ := walFile.FirstIndex(); index == 0 {
+		// simple way to test if the walfile is freshly initialized, meaning no existing file was found
+		log.Printf("I! wal file not found for plugin %s, this can safely be ignored if this is the first instance of this plugin", id)
+	}
+
 	buf := &DiskBuffer{
 		BufferStats: stats,
 		file:        walFile,

--- a/models/buffer_disk.go
+++ b/models/buffer_disk.go
@@ -33,7 +33,7 @@ type DiskBuffer struct {
 	isEmpty bool
 }
 
-func NewDiskBuffer(id, path string, stats BufferStats) (*DiskBuffer, error) {
+func NewDiskBuffer(name, id, path string, stats BufferStats) (*DiskBuffer, error) {
 	filePath := filepath.Join(path, id)
 	walFile, err := wal.Open(filePath, nil)
 	if err != nil {
@@ -42,7 +42,8 @@ func NewDiskBuffer(id, path string, stats BufferStats) (*DiskBuffer, error) {
 	//nolint:errcheck // cannot error here
 	if index, _ := walFile.FirstIndex(); index == 0 {
 		// simple way to test if the walfile is freshly initialized, meaning no existing file was found
-		log.Printf("I! wal file not found for plugin %s, this can safely be ignored if this is the first instance of this plugin", id)
+		log.Printf("I! WAL file not found for plugin outputs.%s (%s), "+
+			"this can safely be ignored if you added this plugin instance for the first time", name, id)
 	}
 
 	buf := &DiskBuffer{

--- a/models/buffer_disk_test.go
+++ b/models/buffer_disk_test.go
@@ -22,7 +22,7 @@ func newTestDiskBuffer(t testing.TB) Buffer {
 
 func newTestDiskBufferWithPath(t testing.TB, name string, path string) Buffer {
 	t.Helper()
-	buf, err := NewBuffer(name, name, "", 0, "disk", path)
+	buf, err := NewBuffer(name, "123", "", 0, "disk", path)
 	require.NoError(t, err)
 	buf.Stats().MetricsAdded.Set(0)
 	buf.Stats().MetricsWritten.Set(0)

--- a/models/buffer_disk_test.go
+++ b/models/buffer_disk_test.go
@@ -22,7 +22,7 @@ func newTestDiskBuffer(t testing.TB) Buffer {
 
 func newTestDiskBufferWithPath(t testing.TB, name string, path string) Buffer {
 	t.Helper()
-	buf, err := NewBuffer(name, "", 0, "disk", path)
+	buf, err := NewBuffer(name, name, "", 0, "disk", path)
 	require.NoError(t, err)
 	buf.Stats().MetricsAdded.Set(0)
 	buf.Stats().MetricsWritten.Set(0)

--- a/models/buffer_disk_test.go
+++ b/models/buffer_disk_test.go
@@ -45,6 +45,7 @@ func TestBuffer_RetainsTrackingInformation(t *testing.T) {
 func TestBuffer_TrackingDroppedFromOldWal(t *testing.T) {
 	path, err := os.MkdirTemp("", "*-buffer-test")
 	require.NoError(t, err)
+	path = filepath.Join(path, "123")
 	walfile, err := wal.Open(path, nil)
 	require.NoError(t, err)
 

--- a/models/buffer_mem_test.go
+++ b/models/buffer_mem_test.go
@@ -8,7 +8,7 @@ import (
 
 func newTestMemoryBuffer(t testing.TB, capacity int) Buffer {
 	t.Helper()
-	buf, err := NewBuffer("test", "", capacity, "memory", "")
+	buf, err := NewBuffer("test", "test", "", capacity, "memory", "")
 	require.NoError(t, err)
 	buf.Stats().MetricsAdded.Set(0)
 	buf.Stats().MetricsWritten.Set(0)

--- a/models/buffer_mem_test.go
+++ b/models/buffer_mem_test.go
@@ -8,7 +8,7 @@ import (
 
 func newTestMemoryBuffer(t testing.TB, capacity int) Buffer {
 	t.Helper()
-	buf, err := NewBuffer("test", "test", "", capacity, "memory", "")
+	buf, err := NewBuffer("test", "123", "", capacity, "memory", "")
 	require.NoError(t, err)
 	buf.Stats().MetricsAdded.Set(0)
 	buf.Stats().MetricsWritten.Set(0)

--- a/models/buffer_suite_test.go
+++ b/models/buffer_suite_test.go
@@ -84,7 +84,7 @@ func MetricTime(sec int64) telegraf.Metric {
 
 func (s *BufferSuiteTest) newTestBuffer(capacity int) Buffer {
 	s.T().Helper()
-	buf, err := NewBuffer("test", "", capacity, s.bufferType, s.bufferPath)
+	buf, err := NewBuffer("test", "test", "", capacity, s.bufferType, s.bufferPath)
 	s.Require().NoError(err)
 	buf.Stats().MetricsAdded.Set(0)
 	buf.Stats().MetricsWritten.Set(0)

--- a/models/buffer_suite_test.go
+++ b/models/buffer_suite_test.go
@@ -84,7 +84,7 @@ func MetricTime(sec int64) telegraf.Metric {
 
 func (s *BufferSuiteTest) newTestBuffer(capacity int) Buffer {
 	s.T().Helper()
-	buf, err := NewBuffer("test", "test", "", capacity, s.bufferType, s.bufferPath)
+	buf, err := NewBuffer("test", "123", "", capacity, s.bufferType, s.bufferPath)
 	s.Require().NoError(err)
 	buf.Stats().MetricsAdded.Set(0)
 	buf.Stats().MetricsWritten.Set(0)

--- a/models/running_output.go
+++ b/models/running_output.go
@@ -104,7 +104,7 @@ func NewRunningOutput(
 		batchSize = DefaultMetricBatchSize
 	}
 
-	b, err := NewBuffer(config.Name, config.Alias, bufferLimit, config.BufferStrategy, config.BufferDirectory)
+	b, err := NewBuffer(config.ID, config.Name, config.Alias, bufferLimit, config.BufferStrategy, config.BufferDirectory)
 	if err != nil {
 		panic(err)
 	}

--- a/models/running_output.go
+++ b/models/running_output.go
@@ -104,7 +104,7 @@ func NewRunningOutput(
 		batchSize = DefaultMetricBatchSize
 	}
 
-	b, err := NewBuffer(config.ID, config.Name, config.Alias, bufferLimit, config.BufferStrategy, config.BufferDirectory)
+	b, err := NewBuffer(config.Name, config.ID, config.Alias, bufferLimit, config.BufferStrategy, config.BufferDirectory)
 	if err != nil {
 		panic(err)
 	}

